src/hotspot/share/services/mallocTracker.cpp

```diff
@@ -103,15 +106,102 @@ void MallocMemorySummary::initialize() {
   ::new ((void*)_snapshot)MallocMemorySnapshot();
 }
 
-void MallocHeader::release() const {
+void MallocHeader::mark_block_as_dead() {
+  _canary = _header_canary_dead_mark;
+  NOT_LP64(_alt_canary = _header_alt_canary_dead_mark);
+  set_footer_byte(_footer_canary_dead_mark);
+}
+
+void MallocHeader::release() {
   // Tracking already shutdown, no housekeeping is needed anymore
   if (MemTracker::tracking_level() <= NMT_minimal) return;
 
+  check_block_integrity();
+
   MallocMemorySummary::record_free(size(), flags());
   MallocMemorySummary::record_free_malloc_header(sizeof(MallocHeader));
   if (MemTracker::tracking_level() == NMT_detail) {
     MallocSiteTable::deallocation_at(size(), _bucket_idx, _pos_idx);
   }
+
+  mark_block_as_dead();
+}
+
+void MallocHeader::print_block_on_error(outputStream* st, address bad_address) const {
+  st->print_cr("NMT Block at " PTR_FORMAT ", corruption at: " PTR_FORMAT ": ",
+               p2i(this), p2i(bad_address));
+  address from = align_down((address)this, sizeof(void*)) - 8;
+  address to = from + 64;
+  // Note: print_hex_dump uses SafeFetch, so it should be able to handle unmapped memory.
+  os::print_hex_dump(st, from, to, 1);
+  assert(bad_address >= from, "sanity");
+  // if the corruption is in the block body of in the footer, print out that part too
+  // unless it has been part of the first hexdump
+  address from2 = align_down(bad_address, sizeof(void*)) - 8;
+  from2 = MAX2(to, from2);
+  address to2 = from2 + 96;
+  if (to2 > to) {
+    if (from2 > to) {
+      st->print_cr("...");
+    }
+    os::print_hex_dump(st, from2, to2, 1);
+  }
+}
+
+// Check block integrity. If block is broken, print out a report
+// to tty (optionally with hex dump surrounding the broken block),
+// then trigger a fatal error.
+void MallocHeader::check_block_integrity() const {
+
+  // Note: if you modify the error messages here, make sure you
+  // adapt the associated gtests too.
+
+  // Weed out obviously wrong block addresses of NULL or very low
+  // values. Note that we should not call this for ::free(NULL),
+  // which should be handled by os::free() above us.
+  if (((size_t)p2i(this)) < K) {
+    fatal("Block at " PTR_FORMAT ": invalid block address", p2i(this));
+  }
+
+  // From here on we assume the block pointer to be valid. We could
+  //  use SafeFetch but since this is a hot path we don't. If we are
+  //  wrong, we will crash when accessing the canary, which hopefully
+  //  generates distinct crash report.
+
+  // Also weed out unaligned addresses. Note that the alignment requirements
+  // we check here are the bare minimum of what we know will malloc() give us
+  // (which is 64-bit even on 32-bit platforms).
+  if (!is_aligned(this, sizeof(uint64_t))) {
```

Comment:
We know that the alignment has to be *at least* 64-bit since we know we have 64-bit inbuilt types on both 64-bit and 32-bit platforms (`uint64_t`). From experience, I know it is probably more, 16 or 32 bytes. This makes sense since there exist scalar data types larger than 64-bit.

But this code tests the *minimal* necessary alignment only and I wanted to prevent false positives. Let's say, in case we happen to run with a libc whose malloc implementation returns only 64-bit aligned pointers. That also could happen if someone put a weird malloc() implementation below us (malloc hooks or LD_PRELOAD). I think the assumption that everything malloc() returns is at least 64-bit aligned is pretty safe though.

Ideally, we would have a clear definition of malloc alignment somewhere in `globalDefinitions.hpp`. In hotspot, there are a range of places where such alignment is implicitly assumed. The NMT header size, for instance, or metaspace allocation size, hotspot arena allocation alignment etc. Basically, everywhere where one either marshalls malloc'ed blocks or implements some sort of general purpose allocator. C++ has `std::max_align_t` for that. Maybe we could use that one. But that's a topic for another RFE.

I try to improve the comment.

Reply:
> Where does this information come from? As far as I can see, the man-page of `malloc()` only mentions:
> 
> > "malloc returns a pointer which is suitably aligned for any built-in  type"
> 
> Why is this 64 bit on a 32-bit platform?

NMT always assumes (from experiments on various platforms) that malloc memory is 2-machine-word aligned, so it is 64-bit align on a 32-bit platform.